### PR TITLE
Fix/chainlight remediations

### PR DESCRIPTION
**Pull request by livingrockrises**
# Summary

Related Issue: # 
https://github.com/chainlight-io/2024-01-biconomy-audit-issues-client/issues/7
https://github.com/chainlight-io/2024-01-biconomy-audit-issues-client/issues/5
https://github.com/chainlight-io/2024-01-biconomy-audit-issues-client/issues/6
https://github.com/chainlight-io/2024-01-biconomy-audit-issues-client/issues/7

## Change Type

- [ ] Bug Fix
- [ ] Refactor
- [ ] New Feature
- [ ] Breaking Change
- [ ] Documentation Update
- [ ] Performance Improvement
- [ ] Other

# Checklist

- [ ] My code follows this project's style guidelines
- [ ] I've reviewed my own code
- [ ] I've added comments for any hard-to-understand areas
- [ ] I've updated the documentation if necessary
- [ ] My changes generate no new warnings
- [ ] I've added tests that prove my fix is effective or my feature works
- [ ] All unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published

# Additional Information

# Branch Naming

### CODE REVIEW

**Comment by livingrockrises** [comment]
> 1. here

can't open this link
https://data.chain.link/_next/data/_iNK4_aT-_RWKGZ3A8gi7/feeds.json

**Comment by livingrockrises** [comment]
> 2\. here

do you mean changing this from storage to memory?

TokenInfo storage tokenInfo = tokensInfo[token];

**Comment by livingrockrises** [comment]
> There is a flaw in MathLib.sol's min function. (It returns the max value rather than the min value.) Change the gt() part in minuint256() and minuint32() as shown below.

Are you sure about this, it does break my test cases.

**Comment by livingrockrises** [comment]
> > There is a flaw in MathLib.sol's min function. (It returns the max value rather than the min value.) Change the gt() part in minuint256() and minuint32() as shown below.
> 
> Are you sure about this, it does break my test cases.

ok I just confirmed in foundry, will check the tests

**Comment by livingrockrises** [comment]
> > [#81 (comment)](https://github.com/bcnmy/biconomy-paymasters/pull/81#issuecomment-1939362513)
> 
> Sorry. We didn't know that the link was temporary. We are uploading the downloaded version. [feeds.json](https://github.com/bcnmy/biconomy-paymasters/files/14257441/feeds.json)
> 
> > [#81 (comment)](https://github.com/bcnmy/biconomy-paymasters/pull/81#issuecomment-1939373807)
> 
> Yes.
> 
> > [#81 (comment)](https://github.com/bcnmy/biconomy-paymasters/pull/81#issuecomment-1939408304)
> 
> Yes.

Thanks. For the deployment scripts will try to fin heartbeat for all supported oracles across supported chains. What value do you recommend for a buffer?

btw, all the changes have been made. can you re-review and approve the PR please

File: contracts/interfaces/paymasters/IBiconomyTokenPaymaster.sol [hunk 1]
```diff
@@ -32,6 +32,8 @@ interface IBiconomyTokenPaymaster {
      */
     event FeeReceiverChanged(address indexed _oldfeeReceiver, address indexed _newfeeReceiver, address indexed _actor);
 
+    event TokensWithdrawn(address indexed _token, address indexed _to, uint256 indexed _amount, address actor);
```
Comment (Aboudjem) [inline contracts/interfaces/paymasters/IBiconomyTokenPaymaster.sol:35]: Not sure `_amount` should be `indexed`

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 2]
```diff
@@ -425,12 +380,90 @@ contract BiconomyTokenPaymaster is
...
+        // This model assumes irrespective of priceSource exchangeRate is always sent from outside
+        // for below checks you would either need maxCost or some exchangeRate
+
+        uint256 btpmRequiredPrefund = _getRequiredPrefund(userOp);
+
+        uint256 tokenRequiredPreFund = (btpmRequiredPrefund * exchangeRate) / 10 ** 18;
+        require(priceMarkup <= 2e6, "BTPM: price markup percentage too high");
+        require(
+            IERC20(feeToken).balanceOf(account) >= ((tokenRequiredPreFund * priceMarkup) / PRICE_DENOMINATOR),
+            "BTPM: account does not have enough token balance"
```
Comment (Aboudjem) [inline contracts/token/BiconomyTokenPaymaster.sol:456]: To fit it into 32 bytes, we can use the following require msg: `BTPM: Insufficient token balance.`

File: contracts/token/oracles/OracleAggregator.sol [hunk 3]
```diff
@@ -27,8 +28,9 @@ abstract contract OracleAggregator is Ownable, IOracleAggregator {
         if (decimals1 != decimals2) revert MismatchInBaseAndQuoteDecimals();
         tokensInfo[token].tokenOracle = tokenOracle;
         tokensInfo[token].nativeOracle = nativeOracle;
-        tokensInfo[token].tokenDecimals = tokenDecimals;
+        tokensInfo[token].tokenDecimals = ERC20(token).decimals();
         tokensInfo[token].isDerivedFeed = isDerivedFeed;
+        tokensInfo[token].priceUpdateThreshold = priceUpdateThreshold;
```
Comment (Aboudjem) [inline contracts/token/oracles/OracleAggregator.sol:33]: Current function `setTokenOracle` updates TokenInfo in tokensInfo map separately. 

I'll suggest to pass TokenInfo directly and update map once. Saves SSTORE operation
```
tokensInfo[token] = TokenInfo({
    tokenDecimals: ERC20(token).decimals(),
    priceUpdateThreshold,
    tokenOracle,
    nativeOracle,
    isDerivedFeed
});
```
This way we use less gas

File: contracts/interfaces/paymasters/IBiconomyTokenPaymaster.sol (hunk 1 above)
Comment (livingrockrises) [reply contracts/interfaces/paymasters/IBiconomyTokenPaymaster.sol:35]: we can index 3 params and I think it's fine. also out of scope of remediations PR

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 2 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol:456]: done

File: contracts/token/oracles/OracleAggregator.sol (hunk 3 above)
Comment (livingrockrises) [reply contracts/token/oracles/OracleAggregator.sol:33]: can we do this as part of another PR? first I will check gas consumptions on forge. this also requires major changes again in scripts

**Review by Aboudjem — APPROVED** [review]

**Pull request merged**